### PR TITLE
feat: add GitHub Codespaces support

### LINKED ISSUE #2216: ✨ Add Github Codespaces configuration
### What are you trying to do?

I'm using Github Codespaces to write contributions for the Dagger repo, but the default Codespaces environment is general purpose and so isn't tailored for this specific repo. It would be useful to have a fixed Codespaces configuration for this repo for repeatable dev environments on any branch.

### Why is this important to you?

I can easily work on this repo without having to make sure I have my dev environment set up correctly. Additionally every branch I work on will always have a version of Dagger compiled from that branch's head.

### How are you currently working around this?

I have to create new config files for each new branch and make sure I don't commit them if making a pull request.

### PR DESCRIPTION
**Pull request by berryp**
Signed-off-by: Berry Phillips <berryphillips@gmail.com>

Closes #2216

The configs are auto-generated by Visual Studio Code and I have added additional configuration to support the basic needs of this repository. I added a section in the FAQ for this. I don't know if that's the best place or if there should be a dedicated doc for this 🤷🏻‍♂️

## Testing

Spin up a new Codespace from my branch and ensure the following:
* Visual Studio Code loads without errors or warnings
* The `dagger` and `cue` commands are available in the terminal and are the correct versions (`dagger` should be compiled from the branch source and `cue` should be the same version as specified in `go.mod`)
* `.cue` files have syntax highlighting and can be auto formatted
* You can run dagger commands that interact with buildkit

### CODE REVIEW

File: docs/guides/1227-contributing.md [hunk 1]
```diff
@@ -172,3 +172,13 @@ git commit --amend -s
 # Force push the new commit to re-run all GitHub Actions jobs:
 git push -f mybranch
 ```
+
+### Can I use a remote development environment?
+
+Yes! The Dagger repository has Github Codespaces configuration included to help help you get started contributing directly from GitHub.
```
Comment (TomChv) [inline docs/guides/1227-contributing.md:178]: - Embed a link to Dagger repository 
- Github -> GitHub
- You repeated `help` 2 times
- To help you contributing directly from GitHub

**Review by TomChv — COMMENTED** [review]
That's an amazing contribution! Thank you 🚀

I left a little comment for now but I'll take a closer look Tuesday.

Just a quick comment : do not forget to also prefix your commit with ˋfeat` so it will be correctly handle by the Changelog

And last thing, is there a way to automatically test .devcontainer? I saw you manually tested but we should find a way to verify with an action or something

**Comment by berryp** [comment]
Thanks for the comments, I'll get those fixed.

> Just a quick comment : do not forget to also prefix your commit with ˋfeat` so it will be correctly handle by the Changelog

Looks like this is missing from the live contribution docs for [commit messages](https://docs.dagger.io/1227/contributing#commit-messages).

> And last thing, is there a way to automatically test .devcontainer? I saw you manually tested but we should find a way to verify with an action or something

One thing we can do is to build and run the container that it uses (there is a Dockerfile), which is pretty heavy but if that test only runs against main then that shouldn't be an issue. I'll look into further possibilities.

**Comment by berryp** [comment]
Added CI job to Build the Codespaces image and run the post-install script. Tried to use a schema validator for the config file `devcontainer.json` but it's not possible with the [schema MS generates](https://github.com/microsoft/vscode/tree/main/extensions/configuration-editing/schemas).

**Comment by gerhard** [comment]
Did you manage to try this out @helderco? I am curious to try it out myself.

**Comment by helderco** [comment]
> Did you manage to try this out @helderco? I am curious to try it out myself.

Not yet, but I want to!

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,52 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the README at:
+// https://github.com/microsoft/vscode-dev-containers/tree/v0.224.3/containers/go
+{
+	"name": "Go",
```
Comment (marcosnils) [inline .devcontainer/devcontainer.json:4]: Shouldn't name be Dagger here?

Comment (berryp) [reply .devcontainer/devcontainer.json]: Fixed.

**Comment by marcosnils** [comment]
:wave: just tried this and it works perfect. I think we should definitely merge it!
image

@berryp what a coincidence that we are  both working on it this sunday after a few days of being stale :raised_hands: . Thx for fixing the conflicts and apologize for taking some time to merge.

**Review by marcosnils — APPROVED** [review]

**Comment by marcosnils** [comment]
@helderco @gerhard @jpadams this LGTM! Lets merge as soon as someone else approves :pray:

**Review by helderco — APPROVED** [review]

**Pull request merged**